extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java

```diff
@@ -41,19 +45,30 @@ private void doHandle(RoutingContext ctx) {
             Arc.container().instance(CurrentIdentityAssociation.class).get().setIdentity(user.getSecurityIdentity());
         }
         SmallRyeHealthReporter reporter = Arc.container().instance(SmallRyeHealthReporter.class).get();
-        SmallRyeHealth health = getHealth(reporter, ctx);
-        HttpServerResponse resp = ctx.response();
-        if (health.isDown()) {
-            resp.setStatusCode(503);
-        }
-        resp.headers().set(HttpHeaders.CONTENT_TYPE, "application/json; charset=UTF-8");
-        Buffer buffer = Buffer.buffer(256); // this size seems to cover the basic health checks
-        try (BufferOutputStream outputStream = new BufferOutputStream(buffer);) {
-            reporter.reportHealth(outputStream, health);
-            resp.end(buffer);
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
+        getHealth(reporter, ctx).subscribe().with(health -> {
+            HttpServerResponse resp = ctx.response();
+            if (health.isDown()) {
+                resp.setStatusCode(503);
+            }
+            resp.headers().set(HttpHeaders.CONTENT_TYPE, "application/json; charset=UTF-8");
+            Buffer buffer = Buffer.buffer(256); // this size seems to cover the basic health checks
+            try (BufferOutputStream outputStream = new BufferOutputStream(buffer)) {
+                reporter.reportHealth(outputStream, health);
+            } catch (Throwable e) {
+                log.error("Unable to prepare health response message", e);
+                resp.setStatusCode(500);
+                resp.end();
```

Comment:
Move it in the finally group and avoid the ! isEnded.

Reply:
I've pushed one more change, which should improve this: The buffer is only sent if `reportHealth()` has completed without errors, and in case of errors there is an empty 500 response.